### PR TITLE
ftests: let systemd create the test slice

### PR DESCRIPTION
**Pull request by gkennedy12**
It is not necessary to create the test slice as systemd creates the slice if it does not exist. In addition, set the _runtime_ option in the sudo tests so that the settings are temporary (same as "systemctl --runtime").

### CODE REVIEW

File: adaptived/tests/ftests/1000-sudo-effect-sd_bus_setting_set_int.json [hunk 1]
```diff
@@ -10,6 +10,16 @@
 				}
 			],
 			"effects": [
+				{
+					"name": "sd_bus_setting",
```
Comment (drakenclimber) [inline adaptived/tests/ftests/1000-sudo-effect-sd_bus_setting_set_int.json:14]: Is this effect needed?  I think `MemoryMax` is set to `infinity` by default.

Or are you using this to test the set and validate logic?

Comment (gkennedy12) [reply adaptived/tests/ftests/1000-sudo-effect-sd_bus_setting_set_int.json:14]: I am finding that if the test is run more than once, the last value set is what it starts with not _infinity_. Not sure what is "remembering" the last value.

Comment (gkennedy12) [reply adaptived/tests/ftests/1000-sudo-effect-sd_bus_setting_set_int.json:14]: We are modifying MemoryMax. Added this debug to show the issue:
```
diff --git a/adaptived/src/utils/sd_bus_utils.c b/adaptived/src/utils/sd_bus_utils.c
index 1896ff5..c401be9 100644
--- a/adaptived/src/utils/sd_bus_utils.c
+++ b/adaptived/src/utils/sd_bus_utils.c
@@ -368,9 +368,25 @@ API int adaptived_sd_bus_set_ll(const char * const target, const char * const pr
        val.type = ADAPTIVED_CGVAL_LONG_LONG;
        val.value.ll_value = value;
 
+       {
+               long long ll_value2 = 0;
+
+fprintf(stderr, "\nXXX PRE:\n");
+               system("cat /sys/fs/cgroup/sudo1000.slice/memory.max");
+               ret = adaptived_sd_bus_get_ll(target, property, &ll_value2);
+fprintf(stderr, "XXX adaptived_sd_bus_set_ll: First get adaptived_sd_bus_get_ll() target=%s, property=%s, ll_value2=%lld, ret=%d\n",
+    target, property, ll_value2, ret);
+fprintf(stderr, "XXX POST:\n");
+               system("cat /sys/fs/cgroup/sudo1000.slice/memory.max");
+fprintf(stderr, "\n");
+       }
        ret = set_property(target, property, &val);
        if (ret < 0)
                return ret;
+fprintf(stderr, "XXX POST set_property():\n");
+               system("ls /sys/fs/cgroup/sudo1000.slice/memory.max");
+               system("cat /sys/fs/cgroup/sudo1000.slice/memory.max");
+fprintf(stderr, "\n");
 
        if (ret == 0 && (flags & ADAPTIVED_CGROUP_FLAGS_VALIDATE)) {
                long long validate_value;
```
And here's the output:
```
sudo1000 .
Debug: Initializing cause always
Debug: Initializing effect sd_bus_setting
Info: Failed to find key pre_set_from
Debug: Parsed cgroup value: ll_value = 89997312
Info: Failed to find key limit
Debug: sd_bus setting: validate = 1
Debug: adaptived_loop: Debug mode. Skip running as daemon.
Debug: Running rule Set an integer cgroup setting via sd_bus
Debug: always triggered
Debug: Running effect sd_bus_setting

XXX PRE:
cat: /sys/fs/cgroup/sudo1000.slice/memory.max: No such file or directory
Debug: get_property: path=/org/freedesktop/systemd1/unit/sudo1000_2eslice, interface=org.freedesktop.systemd1.Slice, prop=MemoryMax, type=t
Debug: get_property: type=t, r=1, val=89997312
XXX adaptived_sd_bus_set_ll: First get adaptived_sd_bus_get_ll() target=sudo1000.slice, property=MemoryMax, ll_value2=89997312, ret=0
XXX POST:
cat: /sys/fs/cgroup/sudo1000.slice/memory.max: No such file or directory

Debug: set_property: name=sudo1000.slice, property=MemoryMax
Debug: get_type_from_string: s=sudo1000.slice
Debug: property_assignment: t=0, property=MemoryMax
Debug: update_property: property=MemoryMax, value->value.ll_value=89997312
XXX POST set_property():
/sys/fs/cgroup/sudo1000.slice/memory.max
89997312

Debug: get_property: path=/org/freedesktop/systemd1/unit/sudo1000_2eslice, interface=org.freedesktop.systemd1.Slice, prop=MemoryMax, type=t
Debug: get_property: type=t, r=1, val=89997312
Debug: adaptived_sd_bus_set_ll: validate success.
Debug: adaptived main loop exceeded max loops
Debug: Cleaning up rule Set an integer cgroup setting via sd_bus
Debug: Cleaning up cause always
Debug: Cleaning up effect sd_bus_setting
Debug: stop_transient: sudo /bin/systemctl stop sudo1000.slice > /dev/null 2>&1, ret=0
```
At first /sys/fs/cgroup/sudo1000.slice/memory.max does not exist, but the MemoryMax property does from its prior setting.

**Comment by gkennedy12** [comment]
rmdir of the slice has no effect on the persistent storage at:
```
/etc/systemd/system.control/sudo1000.slice.d/50-MemoryMax.conf
```
Not sure how to get rid of the persistent storage. Tried: systemctl daemon-reload, systemctl daemon-reexec, systemctl reset-failed

**Comment by drakenclimber** [comment]
> rmdir of the slice has no effect on the persistent storage at:
> 
> ```
> /etc/systemd/system.control/sudo1000.slice.d/50-MemoryMax.conf
> ```
> 
> Not sure how to get rid of the persistent storage. Tried: systemctl stop sudo1000.slice, systemctl daemon-reload, systemctl daemon-reexec, systemctl reset-failed

Bummer.  Then I guess your method of explicitly setting MemoryMax to a known value is probably the best.  A comment explaining why we're doing this would be good though.

**Comment by gkennedy12** [comment]
Is it verboten to just remove /etc/systemd/system.control/foo.slice.d ? I can't find anything with google on removing persistent systemd data.

**Comment by drakenclimber** [comment]
> Is it verboten to just remove /etc/systemd/system.control/foo.slice.d ? I can't find anything with google on removing persistent systemd data.

Good question.  I honestly don't know

**Comment by gkennedy12** [comment]
The property is not persistent if the --runtime switch is used with set-property.
```
systemctl set-property --runtime foo.slice MemoryMax=102400000
```
I think when the process goes away, this goes away (the persistent storage):
```
/run/systemd/system.control/foo.slice.d/50-MemoryMax.conf
```
Trying to figure out how work _--runtime_ into the sd_bus util commands.

File: adaptived/src/effects/sd_bus_setting.c [hunk 2]
```diff
@@ -332,8 +354,14 @@ int sd_bus_setting_main(struct adaptived_effect * const eff)
 			ret = adaptived_get_meminfo_field(PROC_MEMINFO, "MemTotal", &ll_value);
 			if (ret)
 				return ret;
+
+			cgflags |= ADAPTIVED_CGROUP_FLAGS_VALIDATE;
```
Comment (drakenclimber) [inline adaptived/src/effects/sd_bus_setting.c:358]: Should this be
```
cgflags = ADAPTIVED_CGROUP_FLAGS_VALIDATE;
```

What would be the use case where we want to `or` in previous values in cgflags?

Comment (drakenclimber) [reply adaptived/src/effects/sd_bus_setting.c:358]: As currently implemented, it probably doesn't matter whether we use `=` or `|=`, but I'm wondering if in the future we could get bitten by erroneously or'ing in old data.

**Review by drakenclimber — APPROVED** [review]
I had one nitpicky comment.  Also, I was going to ask you to add `runtime` to the documentation for the systemd sd-bus effect, but we don't have any documentation for it.  In another pull request, could you please add documentation around this effect?  Thanks!

File: adaptived/src/effects/sd_bus_setting.c (hunk 2 above)
Comment (gkennedy12) [reply adaptived/src/effects/sd_bus_setting.c:358]: Made the change.

**Comment by gkennedy12** [comment]
Will add doc for sd-bus effect and runtime.

**Review by drakenclimber — APPROVED** [review]

**Comment by drakenclimber** [comment]
Looks good.  Thanks, @gkennedy12

**Comment by gkennedy12** [comment]
Not quite sure how to handle the case where there is left over persistent data from previous runs without the _runtime_ option. The systemd persistent slice data ends up here if the _runtime_ option has not been set: /etc/systemd/system.control . If the tests had never been run before and now the _runtime_ option is set, there will be no persistent data in /etc/systemd/system.control. We have run the tests without the _runtime_ option, so it is probably likely that there is persistent data that needs to be cleared out. This should only be a problem for us & Jeff. To start off clean run:
```
rm -rf /etc/systemd/system.control
```

**Comment by sidkumar99** [comment]
adaptivemm binary is not affected
Acked-by: Sidhartha Kumar <sidhartha.kumar@oracle.com>

**Comment by drakenclimber** [comment]
Merged in pull request https://github.com/oracle/adaptivemm/pull/29.  Thanks, @gkennedy12 and @sidkumar99

**Pull request closed without merging**